File: day-12/day-12.py
import re

with open('../data/12-input.txt', 'r') as f:
    data = f.read().splitlines()

state_input = re.findall('initial state: (.*)', data[0])[0]
state = {i:st for i, st in enumerate(state_input)}
init_state = state.copy()

rule_lines = data[2:]
rules = {line[:5]:line[-1] for line in rule_lines}

def get_situation(pot, state):
    pots = [pot - 2, pot - 1, pot, pot + 1, pot + 2]
    situation = ''.join([state.get(p, '.') for p in pots])
    return situation

def run_sim(n_gen, state):
    for gen in range(n_gen):

        min_pot = min(state.keys())
        max_pot = max(state.keys())
        state[min_pot - 2] = '.'
        state[min_pot - 1] = '.'
        state[max_pot + 1] = '.'
        state[max_pot + 2] = '.'

        state = {pot:rules[get_situation(pot, state)] for pot in state.keys()}

    total = 0
    for pot, value in state.items():
        if value == '#':
            total += pot

    return total

total_20 = run_sim(20, state)

print("::: PART A")
print(f"Total after 20 generations: {total_20}\n")

#==============================================================================

print("::: PART B")
print(f"Total after 50B generations: {50_000_000_000 * 800}")

File: day-12/day-12.scala
import scala.io.Source

val lines = Source.fromFile("../data/12-input.txt").getLines.toArray

var initState = lines(0).split(" ").last
    .zipWithIndex
    .map(t => t._2 -> t._1.toString)
    .toMap

val rules = lines
    .drop(2)
    .map(s => s.split(" => "))
    .map(t => t(0) -> t(1))
    .toMap

def getLayout(n: Int) = ((n - 2) to (n + 2)).map(state.getOrElse(_, ".")).mkString

def simGeneration(state: Map[Int, String]) = {
    val minPot = state.keys.min
    val maxPot = state.keys.max
    
    val stateWider = state ++ List(
        (minPot - 2 -> "."),
        (minPot - 1 -> "."),
        (maxPot + 1 -> "."),
        (maxPot + 2 -> ".")
    )

    stateWider
        .map(t => t._1 -> getLayout(t._1))
        .map(t => t._1 -> rules(t._2))
}

def countPlants(state: Map[Int, String]) = state.map(t => if (t._2 == "#") t._1 else 0).sum

var state = initState
for (i <- 1 to 20) state = simGeneration(state)

println(countPlants(state))
